File: bridge-node/src/routes/shows.ts
/**
 * Shows 路由
 * 剧集的季和集列表
 */

import { Hono } from 'hono';
import { config } from '../config.ts';
import { generateServerId, toFnosGuid, toJellyfinId, registerItemType } from '../mappers/id.ts';
import { mapPlayListItemToDto, type BaseItemDto } from '../mappers/item.ts';
import { requireAuth } from '../middleware/auth.ts';
import { fnosGetSeasonList, fnosGetEpisodeList } from '../services/fnos.ts';
import { setImageCache } from '../services/imageCache.ts';
import type { SessionData } from '../services/session.ts';

const shows = new Hono();

const serverId = generateServerId(config.fnosServer);

/**
 * GET /Shows/:seriesId/Seasons
 * 获取剧集的季列表
 */
shows.get('/:seriesId/Seasons', requireAuth(), async (c) => {
  const session = c.get('session') as SessionData;
  const seriesId = c.req.param('seriesId');
  const fnosGuid = toFnosGuid(seriesId);

  if (!fnosGuid) {
    return c.json({ Items: [], TotalRecordCount: 0, StartIndex: 0 });
  }

  try {
    console.log(`[SHOWS] Seasons请求: seriesId=${seriesId}, fnosGuid=${fnosGuid}`);
    const result = await fnosGetSeasonList(session.fnosServer, session.fnosToken, fnosGuid);
    console.log(`[SHOWS] Seasons结果: success=${result.success}, seasons=${result.data?.length || 0}`);
    if (!result.success || !result.data) {
      return c.json({ Items: [], TotalRecordCount: 0, StartIndex: 0 });
    }

    const seasons: BaseItemDto[] = result.data.map(season => {
      const dto: BaseItemDto = {
        Name: season.title || `第 ${season.season_number} 季`,
        ServerId: serverId,
        Id: toJellyfinId(season.guid),
        CanDelete: false,
        CanDownload: false,
        IsFolder: true,
        Type: 'Season',
        IndexNumber: season.season_number,
        SeriesId: seriesId,
        SeriesName: season.tv_title || '',
        ImageTags: season.poster ? { Primary: Buffer.from(season.poster).toString('base64url').slice(0, 16) } : {},
        BackdropImageTags: [],
        LocationType: 'FileSystem',
        ChildCount: season.local_number_of_episodes || season.episode_number || 0,
      };
      // 注册季的类型，供详情路由判断
      registerItemType(season.guid, 'Season');
      // 缓存季的 poster 图片
      if (season.poster) {
        setImageCache(dto.Id, {
          poster: season.poster,
          server: session.fnosServer,
          token: session.fnosToken,
        });
      }
      return dto;
    });

    return c.json({
      Items: seasons,
      TotalRecordCount: seasons.length,
      StartIndex: 0,
    });
  } catch (e: any) {
    console.error('获取季列表失败:', e.message);
    return c.json({ Items: [], TotalRecordCount: 0, StartIndex: 0 });
  }
});

/**
 * GET /Shows/:seriesId/Episodes
 * 获取某季的集列表
 */
shows.get('/:seriesId/Episodes', requireAuth(), async (c) => {
  const session = c.get('session') as SessionData;
  const seriesId = c.req.param('seriesId');
  const seasonId = c.req.query('SeasonId') || c.req.query('seasonId');
  const startIndex = parseInt(c.req.query('StartIndex') || '0', 10);
  const limit = parseInt(c.req.query('Limit') || '100', 10);

  console.log(`[SHOWS] Episodes请求: seriesId=${seriesId}, seasonId=${seasonId}`);

  // episode/list 需要季 guid
  let fnosGuid: string | null = null;
  if (seasonId) {
    fnosGuid = toFnosGuid(seasonId);
  } else {
    // 如果没有 SeasonId，先获取系列的第一个季
    const seriesFnosGuid = toFnosGuid(seriesId);
    if (!seriesFnosGuid) {
      return c.json({ Items: [], TotalRecordCount: 0, StartIndex: startIndex });
    }
    const seasonsResult = await fnosGetSeasonList(session.fnosServer, session.fnosToken, seriesFnosGuid);
    if (seasonsResult.success && seasonsResult.data && seasonsResult.data.length > 0) {
      fnosGuid = seasonsResult.data[0].guid;
    }
  }

  if (!fnosGuid) {
    return c.json({ Items: [], TotalRecordCount: 0, StartIndex: startIndex });
  }

  try {
    const result = await fnosGetEpisodeList(session.fnosServer, session.fnosToken, fnosGuid);
    console.log(`[SHOWS] Episodes结果: fnosGuid=${fnosGuid}, episodes=${result.data?.length || 0}`);
    if (!result.success || !result.data) {
      return c.json({ Items: [], TotalRecordCount: 0, StartIndex: startIndex });
    }

    let episodes = result.data;

    const allItems = episodes.map(ep => {
      const dto = mapPlayListItemToDto(ep, serverId);
      // 缓存集的 poster
      if (ep.poster) {
        setImageCache(dto.Id, {
          poster: ep.poster,
          server: session.fnosServer,
          token: session.fnosToken,
        });
      }
      return dto;
    });
    const paged = allItems.slice(startIndex, startIndex + limit);

    return c.json({
      Items: paged,
      TotalRecordCount: allItems.length,
      StartIndex: startIndex,
    });
  } catch (e: any) {
    console.error('获取集列表失败:', e.message);
    return c.json({ Items: [], TotalRecordCount: 0, StartIndex: startIndex });
  }
});

/**
 * GET /Shows/NextUp
 * 下一集推荐（暂返回空）
 */
shows.get('/NextUp', requireAuth(), (c) => {
  return c.json({ Items: [], TotalRecordCount: 0, StartIndex: 0 });
});

export default shows;

File: bridge-rust/src/routes/shows.rs
/// Shows 路由 — 剧集（季/集）

use axum::{
    extract::{Path, Query, State},
    routing::get,
    Json, Router,
};
use serde::Deserialize;
use serde_json::json;

use crate::config::BridgeConfig;
use crate::mappers::id::*;
use crate::mappers::item::*;
use crate::middleware::auth::require_auth;
use crate::services::fnos::*;
use crate::services::session::SessionData;
use crate::types::jellyfin::{BaseItemDto, ItemsResult};

pub fn router() -> Router<BridgeConfig> {
    Router::new()
        .route(
            "/Shows/{seriesId}/Seasons",
            get(seasons).layer(axum::middleware::from_fn(require_auth)),
        )
        .route(
            "/Shows/{seriesId}/Episodes",
            get(episodes).layer(axum::middleware::from_fn(require_auth)),
        )
        .route(
            "/Shows/NextUp",
            get(next_up).layer(axum::middleware::from_fn(require_auth)),
        )
}

#[derive(Deserialize, Default)]
struct SeasonsQuery {
    #[serde(rename = "UserId")]
    user_id: Option<String>,
}

#[derive(Deserialize, Default)]
struct EpisodesQuery {
    #[serde(rename = "SeasonId")]
    season_id: Option<String>,
    #[serde(rename = "Season")]
    season: Option<i32>,
    #[serde(rename = "UserId")]
    user_id: Option<String>,
}

async fn seasons(
    State(config): State<BridgeConfig>,
    Path(series_id): Path<String>,
    Query(_query): Query<SeasonsQuery>,
    req: axum::extract::Request,
) -> axum::response::Response {
    use axum::response::IntoResponse;
    use axum::http::StatusCode;

    let session = match req.extensions().get::<SessionData>() {
        Some(s) => s.clone(),
        None => return (StatusCode::UNAUTHORIZED, Json(json!({"error":"Unauthorized"}))).into_response(),
    };

    let server_id = generate_server_id(&config.fnos_server);
    let fnos_guid = match to_fnos_guid(&series_id) {
        Some(g) => g,
        None => return Json(ItemsResult { items: vec![], total_record_count: 0, start_index: 0 }).into_response(),
    };

    let result = fnos_get_season_list(
        &session.fnos_server,
        &session.fnos_token,
        &fnos_guid,
        &config,
    )
    .await;

    if !result.success || result.data.is_none() {
        return Json(ItemsResult { items: vec![], total_record_count: 0, start_index: 0 }).into_response();
    }

    let seasons = result.data.unwrap();
    let items: Vec<BaseItemDto> = seasons
        .iter()
        .map(|s| {
            register_item_type(&s.guid, "Season");
            let mut dto = map_playlist_item_to_dto(s, &server_id);
            dto.item_type = "Season".into();
            dto.index_number = Some(s.season_number);
            dto.series_id = Some(to_jellyfin_id(&fnos_guid));
            dto.series_name = Some(s.tv_title.clone());
            if dto.name.is_empty() {
                dto.name = format!("第 {} 季", s.season_number);
            }
            dto
        })
        .collect();

    let total = items.len() as i64;
    Json(ItemsResult { items, total_record_count: total, start_index: 0 }).into_response()
}

async fn episodes(
    State(config): State<BridgeConfig>,
    Path(series_id): Path<String>,
    Query(query): Query<EpisodesQuery>,
    req: axum::extract::Request,
) -> axum::response::Response {
    use axum::response::IntoResponse;
    use axum::http::StatusCode;

    let session = match req.extensions().get::<SessionData>() {
        Some(s) => s.clone(),
        None => return (StatusCode::UNAUTHORIZED, Json(json!({"error":"Unauthorized"}))).into_response(),
    };

    let server_id = generate_server_id(&config.fnos_server);

    // 优先用 SeasonId 获取集列表
    let season_guid = query
        .season_id
        .as_ref()
        .and_then(|sid| to_fnos_guid(sid));

    let season_guid = match season_guid {
        Some(g) => g,
        None => {
            // 没有 SeasonId，用 series 的第一季
            let series_guid = match to_fnos_guid(&series_id) {
                Some(g) => g,
                None => return Json(ItemsResult { items: vec![], total_record_count: 0, start_index: 0 }).into_response(),
            };
            let seasons_result = fnos_get_season_list(
                &session.fnos_server,
                &session.fnos_token,
                &series_guid,
                &config,
            )
            .await;
            if !seasons_result.success || seasons_result.data.is_none() {
                return Json(ItemsResult { items: vec![], total_record_count: 0, start_index: 0 }).into_response();
            }
            let seasons = seasons_result.data.unwrap();
            let target_season = query.season.unwrap_or(1);
            match seasons.iter().find(|s| s.season_number == target_season) {
                Some(s) => s.guid.clone(),
                None => match seasons.first() {
                    Some(s) => s.guid.clone(),
                    None => return Json(ItemsResult { items: vec![], total_record_count: 0, start_index: 0 }).into_response(),
                },
            }
        }
    };

    let result = fnos_get_episode_list(
        &session.fnos_server,
        &session.fnos_token,
        &season_guid,
        &config,
    )
    .await;

    if !result.success || result.data.is_none() {
        return Json(ItemsResult { items: vec![], total_record_count: 0, start_index: 0 }).into_response();
    }

    let episodes = result.data.unwrap();
    let items: Vec<BaseItemDto> = episodes
        .iter()
        .map(|ep| {
            let mut dto = map_playlist_item_to_dto(ep, &server_id);
            dto.series_id = Some(to_jellyfin_id(
                &to_fnos_guid(&series_id).unwrap_or_default(),
            ));
            dto
        })
        .collect();

    let total = items.len() as i64;
    Json(ItemsResult { items, total_record_count: total, start_index: 0 }).into_response()
}

async fn next_up() -> Json<ItemsResult> {
    Json(ItemsResult {
        items: vec![],
        total_record_count: 0,
        start_index: 0,
    })
}
